### PR TITLE
[vLLM-ATOM] Support Eagle 3.1 spec decoding

### PR DESCRIPTION
**Pull request by kliuae**
## Motivation
This PR adds EAGLE 3/3.1 speculative decoding support to vLLM-ATOM and its support for Kimi-K2.6.
Supported combinations are MLA target with MLA/MHA draft models. In the case of MLA target with MHA draft, separate KV cache pools are built so that vLLM doesn't attempt to unify the page sizes.

## Test Plan
Target model: `amd/Kimi-K2.6-MXFP4`
Draft model: `lightseekorg/kimi-k2.6-eagle3.1-mla` (MLA), `lightseekorg/kimi-k2.6-eagle3` (MHA)

Server command
```
vllm serve amd/Kimi-K2.6-MXFP4 \
    --host localhost \
    --port 8000 \
    --async-scheduling \
    --trust-remote-code \
    --gpu_memory_utilization 0.8 \
    --kv-cache-dtype fp8 \
    --no-enable-prefix-caching \
    --enable-auto-tool-choice \
    --tool-call-parser kimi_k2 \
    --reasoning-parser kimi_k2  \
    --compilation-config '{"cudagraph_mode": "FULL_AND_PIECEWISE"}' \
    --tensor-parallel-size 8 \
    --speculative-config.method eagle3 \
    --speculative-config.model "$DRAFT_MODEL" \
    --speculative-config.num_speculative_tokens 3
```

lm_eval
```
lm_eval --model local-completions --model_args model=amd/Kimi-K2.6-MXFP4,base_url=http://localhost:8000/v1/completions,num_concurrent=64,max_retries=3,tokenized_requests=False,trust_remote_code=True --tasks gsm8k --num_fewshot 5
```

## Test Result
lm_eval with gsm8k

`lightseekorg/kimi-k2.6-eagle3.1-mla`

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|_  |0.9416|_  |0.0065|
|     |       |strict-match    |     5|exact_match|_  |0.9409|_  |0.0065|

Per-position acceptance rate: 0.800, 0.602, 0.429, Avg Draft acceptance rate: 61.0%

`lightseekorg/kimi-k2.6-eagle3`

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|_  |0.9348|_  |0.0068|
|     |       |strict-match    |     5|exact_match|_  |0.9356|_  |0.0068|

Per-position acceptance rate: 0.819, 0.605, 0.388, Avg Draft acceptance rate: 60.4%

`lightseekorg/kimi-k2.6-eagle3.1-mla`

| Mode | ISL/OSL | Conc | EAGLE | Req/s | delta | Total tok/s | delta | Output tok/s | delta | TTFT mean (ms) | delta | TPOT mean (ms) | delta | ITL mean (ms) | delta |
|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|
| ATOM native | 1k/1k | 64 | spec=3 | 3.726 | - | 7636.8 | - | 3821.0 | - | 869 | - | 13.63 | - | 44.86 | - |
| vLLM plugin | 1k/1k | 64 | spec=3 | 3.634 | -2.5% | 7441.7 | -2.6% | 3720.9 | -2.6% | 869 | +0% | 13.29 | -2.5% | 44.09 | -1.7% |

## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

### CODE REVIEW

File: atom/plugin/vllm/spec_decode_patch.py [hunk 1]
```diff
@@ -4,6 +4,36 @@
...
+    # vLLM's V1 EAGLE proposer SpecDecodeBaseProposer.propose() has an explicit
+    # isinstance() check for native vLLM EAGLE3 model classes before calling
+    # `combine_hidden_states()`. ATOM's vLLM plugin mode provides the same behavior
+    # through the ATOMModelBase wrapper, so patch the type checks to accept the
+    # ATOMModelBase wrapper
+    try:
+        from atom.plugin.vllm.model_wrapper import ATOMModelBase
+        import vllm.v1.spec_decode.llm_base_proposer as llm_base_proposer
+    except Exception:
+        return
```
Comment (zejunchen-zejun) [inline atom/plugin/vllm/spec_decode_patch.py:22]: maybe we need to raise some warning here if import path doesn't exist when upgrading vllm to avoid the silent return

Comment (kliuae) [reply atom/plugin/vllm/spec_decode_patch.py:22]: Good point, added a warning message.

**Review by zejunchen-zejun — DISMISSED** [review]

**Review by yhl-amd — DISMISSED** [review]
LGTM

**Review by zejunchen-zejun — DISMISSED** [review]

**Review by whx-sjtu — DISMISSED** [review]

**Review by zejunchen-zejun — DISMISSED** [review]

File: atom/models/llama.py [hunk 2]
```diff
@@ -514,3 +514,14 @@ def compute_logits(
     ) -> Optional[torch.Tensor]:
         logits = self.lm_head(hidden_states)
         return logits
+
+    def set_aux_hidden_state_layers(self, layers: tuple[int, ...]) -> None:
+        self.model.aux_hidden_state_layers = layers
+
+    def get_eagle3_aux_hidden_state_layers(self) -> tuple[int, ...]:
```
Comment (zejunchen-zejun) [inline atom/models/llama.py]: Hi, @kliuae 
The function `set_aux_hidden_state_layers` and `get_eagle3_aux_hidden_state_layers` should be moved to draft model right? It belongs to draft model.

Comment (kliuae) [reply atom/models/llama.py]: These actually belong to the target model. They are part of the target model's interface telling which of its layers should be used as EAGLE 3 layers. So I think they should stay here in the target model.

Comment (zejunchen-zejun) [reply atom/models/llama.py]: Make sense. The aux helper function should be added for target model. For Llama model, there is no requirement for Llama for a long time, including its spec decoding, so here let's focus on kimi

Comment (kliuae) [reply atom/models/llama.py]: I see, that's a good point. I'll focus this PR on supporting Kimi's EAGLE 3 for now.

**Review by zejunchen-zejun — APPROVED** [review]

**Review by valarLip — APPROVED** [review]
